### DISCUSSION TITLE
Fails to run in docker, can't create UDP port

### DISCUSSION THREAD

**Discussion by livlif2dfullest** [briis/hass-weatherflow2mqtt#150] [Q&A] [answered]
I'm getting the following error when I start my docker compose container:

**ERROR:weatherflow2mqtt.weatherflow_mqtt:Could not start listening to the UDP Socket. Error is: Could not open a local UDP endpoint**

Here is my docker compose setup:
```
  weatherflow2mqtt:
    container_name: weatherflow2mqtt
    environment:
      - TZ=America/Chicago
      - UNIT_SYSTEM=imperial
      - LANGUAGE=en
      - RAPID_WIND_INTERVAL=0
      - DEBUG=False
      - ELEVATION=0
      - WF_HOST=<IP to Hub>
      - WF_PORT=50222
      - MQTT_HOST=mosquitto
      - MQTT_PORT=1883
      #- MQTT_USERNAME=
      #- MQTT_PASSWORD=
      - MQTT_DEBUG=False
      - STATION_ID=<Station Name>
      - STATION_TOKEN=<Station Token>
      - FORECAST_INTERVAL=30
    image: briis/weatherflow2mqtt:latest
    #network_mode: host
    ports:
      - 50222:50222/udp
      #- <Host IP Address>:50222:50222/udp
      #- 0.0.0.0:50222:50222/udp
    restart: unless-stopped
    volumes:
      - /srv/homeautomation/weatherflow2mqtt:/data
```
It works if I uncomment network_mode: host but then it can't access my MQTT since that's running in a container w/o a public port. I don't want to grant access to the host for this container.  Is this possible?  I also tried using the different commented out ports but they produced the same error.

Any guidance would be appreciated.

**Comment by briis** [comment, marked as answer]
The error message is that the WF2MQTT container cannot connect to the UDP port (50222) with the above setup, and that is why `network_mode: host` needs to be on. I then understand that if you do this, the access to the MQTT server is blocked.

I am honestly not a network person and know little about how to configure this setup for both these things to work and not compromising the security setup you want.  May try and ask the question in the Community Forum, here you will get a broader audience: https://community.home-assistant.io/t/weatherflow-to-mqtt-setup/416369

**Reply by livlif2dfullest** [reply]
A little more information, I enabled it to connect to MQTT with network_mode: host, and now I'm seeing the error again.  I'm wondering if it's not a failure of creating the UDP port for the container but a failure to access the UDP port of the hub?  If my IP address is wrong for the hub would the same error be displayed?  I figured out the IP address by process of elimination but I'm still unsure if it's the right IP, is there a better way to figure out the IP of the Tempest hub?

**Reply by josephquigley** [reply]
@livlif2dfullest I know this thread is almost a year old, but I am stuck on this same issue as well. I have confirmed on my dev machine that I can connect to my Weathflow hub via UDP with an alternative python project (such as https://github.com/peted-davis/WeatherFlow_PiConsole), but not `hass-weatherflow2mqtt`. So I believe the issue is container/port related and has nothing to do with the base station.

**Reply by josephquigley** [reply]
I removed the hard-coded IP address and was able to connect to the Weatherflow base station. 🤷

**Reply by livlif2dfullest** [reply]
@josephquigley I got it working a while ago and has been good ever since.  The issue turned out to be my firewall, I had to open up port 50222/udp in order for my weatherflow container to communicate with my Tempest hub.  Here's my config to hopefully help you:
```
  weatherflow2mqtt:
    container_name: weatherflow2mqtt
    environment:
      - TZ=<time zone>
      - UNIT_SYSTEM=imperial
      - LANGUAGE=en
      - RAPID_WIND_INTERVAL=0
      - DEBUG=False
      #https://whatismyelevation.com/ from phone, use calc at bottom of page
      - ELEVATION=<elevation>
      - LATITUDE=<lat>
      - LONGITUDE=<lon>
      - WF_HOST=0.0.0.0
      - WF_PORT=50222
      - MQTT_HOST=mosquitto
      - MQTT_USERNAME=<user>
      - MQTT_PASSWORD=<password>
      - MQTT_PORT=<port>
      - MQTT_DEBUG=False
      - STATION_ID=<id>
      - STATION_TOKEN=<token>
      - FORECAST_INTERVAL=30
    depends_on:
      - mosquitto
    image: briis/weatherflow2mqtt:latest
    ports:
      - 0.0.0.0:50222:50222/udp
    restart: unless-stopped
    volumes:
      - /<path to container configs>/weatherflow2mqtt:/data
 ```

Then on your container machine, add the 50222/udp port to your firewall using:
```
sudo ufw status <- to see which ports are currently open in your firewall
sudo ufw allow 50222/udp <- to open up the UDP port so the hub and the weatherflow container can communicate.
To                         Action      From
--                         ------      ----     
50222/udp                  ALLOW       Anywhere
50222/udp (v6)             ALLOW       Anywhere (v6) 
```
I have other ports opened for other containers, but just listed the ones that were important above.

Hopefully this gets you up and running, let me know :)

**Reply by buzzkc** [reply]
Just to help anyone that may have the Tempest Hub on a different subnet or vlan than where your docker container is running, the udp broadcast is for the same IP subnet as the hub. Either you have to re-broadcast the udp 50222 traffic on the subnet your docker container is on, or connect them on the same networks. I'm running a proxmox server and was able to just add a virtual nic to tagged with my IoT network vlan and connect it to the Iot subnet.
